docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,849 @@
+call the analogue of `oe_verify_report` to verify the data and the Open Enclave
+runtime can use this UUID to determine what plugin verification routine to run.
+
+The plug-in model also exposes a new set of functions that support collaterals.
+
+### Plugin API
+
+Each plugin must implement the functions below:
+
+```C
+/**
+ * Struct that defines the structure of each plugin. Each plugin must
+ * define an UUID for its format and implement the functions in this
+ * struct. Ideally, each plugin should provide a helper function to
+ * create this struct on the behalf of the plugin users.
+ */
+struct oe_attestaton_plugin_t
+{
+    /**
+     * The UUID for the plugin.
+     */
+    uuid_t format_id;
+
+    /**
+     * The function that gets executed when a plugin is registered.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @param[in] config_data An optional pointer to the configuration data.
+     * @param[in] config_data_size The size in bytes of config_data.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*on_register)(
+        oe_attestaton_plugin_t* plugin_context,
+        const void* config_data,
+        size_t config_data_size);
+
+    /**
+     * The function that gets executed when a plugin is unregistered.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*on_unregister)(
+        oe_attestaton_plugin_t* plugin_context);
+
+    /**
+     * Generates the attestation evidence, which is defined as the data
+     * produced by the enclave. The caller may pass in custom claims, which
+     * must be attached to the evidence and then cryptographically signed.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @param[in] custom_claims A buffer to the optional custom claims.
+     * @param[in] custom_claims_size The size in bytes of custom_claims.
+     * @param[out] evidence_buffer An output pointer that will be assigned the
+     * address of the evidence buffer.
+     * @param[out] evidence_buffer_size A pointer that points to the size of the
+     * evidence buffer.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*get_evidence)(
+        oe_attestaton_plugin_t* plugin_context,
+        const uint8_t* custom_claims,
+        size_t custom_claims_size,
+        uint8_t** evidence_buffer,
+        size_t* evidence_buffer_size);
+
+    /**
+     * Frees the generated attestation evidence.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @param[in] evidence_buffer A pointer to the evidence buffer.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*free_evidence)(
+        oe_attestaton_plugin_t* plugin_context,
+        uint8_t* evidence_buffer);
+
+    /**
+     * Generates the attestation collateral, which is defined as the data
```

Comment:
Only in the enclave.

Reply:
For things like the CRL, the Verifier could look it up itself without the enclave having to send it, right?  If the Verifier retrieves it itself, it would get the latest CRL, whereas one sent by the enclave may be out of date.   I think you're saying that this API is for the enclave, but the Verifier might use a separate API to get collateral about the enclave, yes?

Reply:
What I meant to say is, this public function can only be used in the enclave, but in the case where the user does not provide any collaterals, the verifier will get the collaterals via an internal shared function.